### ISSUE TITLE
Discuss: Changing to Hybrid Polymorphism

### ISSUE THREAD

**Issue by samasri** [eclipse/omr#2583]
# Issues with Current Static Polymorphism Implementation

At the moment, existing Eclipse OMR functionality is specialized through extending classes in sub-namespaces and overriding functions via static polymorphism. Although this implementation works, it has some undesirable consequences:

* To ensure that the function from the most derived class is always used, the `self()` function was introduced and required for functional correctness. However, for OMR downstream project consumers (typically language developers) who are new to OMR, the use of `self()` is not obvious, and affects OMR's consumability.
* For language developers using the OMR code, it is not clear to them which functions are expected, recommended, or even necessary to override in order to connect their project to OMR.

# Proposition

Our proposition is to alter the implementation used from purely static polymorphism to hybrid polymorphism:
* Static polymorphism for functions that are intended to be used only within the OMR project namespace.
* Dynamic polymorphism for functions that are intended to be used and extended by client developers in classes outside the OMR project namespace (that is: client project namespace, or language namespace). 
* We also propose adding the macro `OMR_EXTENSION` to the virtualized API functions: `virtual OMR_EXTENSION void extensibleFunction();`. Right now, this macro would be defined to nothing and thus has no use. However, this gives the flexibility for OMR and language developers to change the definition of this macro to things like (1) `__attribute__((annotate("OMR_Extensible")))`, which can then be used by static analysis to do special processing/analysis of these API functions or (2) `final` to prevent further downstream extensions of the function.

# Advantages of the proposed solution

A hybrid solution of static and dynamic polymorphism gives us the following advantages:

* Having a better mechanism to describe the OMR API and extension points (using `virtual` and `OMR_EXTENSION`). In other words, client developers can easily spot functions that are recommended or necessary to override by searching for `OMR_EXTENSION` functions.
* Making the code more approachable by removing the necessity to use `self()` by client developers.
* Having pure virtual functions ensure that client developers implement necessary functions.
* Documenting all API functions in OMR, which has not been done so far.

# Running Benchmarks

* We propose checking for potential performance degradation by running the benchmarks and comparing the run-time before and after the change. Additionally, more run-time profiling can be done down the line.

# Example
Example 1 (practical): I forked Eclipse OMR [here](https://github.com/samasri/omr/) and created a [PR](https://github.com/eclipse/omr/pull/2582) to virtualize a small class `ELFRelocationResolver` in order to show how the process of virtualization would happen.

Example 2 (theoretical - real PR to be submitted soon): Consider 2 functions in `OMR::CodeGenerator`: `lowerTreesPreTreeTopVisit(TR::TreeTop *,vcount_t)` and `trStackMemory()`. The former is overridden in OpenJ9 and is part of the API we would like to highlight for client developers whereas the latter is not. Hence, the virtualization steps in that case would be the following:

0. [This happens only once] Define `OMR_EXTENSION` in [Annotations.hpp](https://github.com/eclipse/omr/blob/master/compiler/infra/Annotations.hpp#L76)
1. Comment `OMR_EXTENSIBLE` in [`OMR::CodeGenerator`](https://github.com/eclipse/omr/blob/master/compiler/codegen/OMRCodeGenerator.hpp#L242), [`OMR::X86::CodeGenerator`](https://github.com/eclipse/omr/blob/master/compiler/x/codegen/OMRCodeGenerator.hpp#L272), [`OMR::X86::AMD64::CodeGenerator`](https://github.com/eclipse/omr/blob/master/compiler/x/amd64/codegen/OMRCodeGenerator.hpp#L53), [`J9::CodeGenerator`](https://github.com/eclipse/openj9/blob/master/runtime/compiler/trj9/codegen/J9CodeGenerator.hpp#L60), and [`TR::CodeGenerator`](https://github.com/eclipse/openj9/blob/master/runtime/compiler/trj9/codegen/CodeGenerator.hpp#L30)
2. Add `virtual OMR_EXTENSION` in the [definition of `OMR::CodeGenerator::lowerTreesPreTreeTopVisit(TR::TreeTop *,vcount_t)`](https://github.com/eclipse/omr/blob/master/compiler/codegen/OMRCodeGenerator.hpp#L292) (but not in [`OMR::CodeGenerator::trStackMemory()`](https://github.com/eclipse/omr/blob/master/compiler/codegen/OMRCodeGenerator.hpp#L273) since it is not part of the API)
3. Remove all `self()` calls of the virtualized function(`lowerTreesPreTreeTopVisit`). In OMR and OpenJ9, `lowerTreesPreTreeTopVisit` is only called once, [here](https://github.com/eclipse/omr/blob/master/compiler/codegen/OMRCodeGenerator.cpp#L370).

**Comment by dibyendumajumdar** [comment]
Hi, 

I am trying to figure out what the extension points are and also what the minimal requirement is, so I understand the pain here and the reason for looking at alternative methods. However I think that:

a) Part of the problem is that while there are docs talking about the current static polymorphism approach in general, there is a lack of detailed information on exactly what the extension points are, and also which ones must be over-ridden, which ones are optional etc. Having more guidance here can help overcome some of the issues.

b) Secondly I can see that the current design allows an implementation to override a lot of stuff - simply by the way headers are looked up. Will the proposed changes here limit this in any way? I think if the changes proposed here were to somehow restrict the current flexibility then it would not be so great. 

c) Thirdly I for one want to work with the core IL generation modules - hence any solution should also allow that and not assume that clients will only access a subset.

I thought it worth putting my thoughts here although my understanding is imperfect right now - hence I do apologise if I have misunderstood something here.

**Comment by samasri** [comment]
Hello Dibyendu,

a) Exactly, since we lack detailed documentation about what are the extension points, this proposal would help fix that by annotating the extension points (or, what we can call API functions) with `OMR_EXTENSION` (although there have been [recent discussions](https://github.com/eclipse/omr/pull/2584#issuecomment-393595121) of changing the name of the annotation to `OMR_API`).

b) Since we will only be changing the way overridden functions are resolved from static to dynamic, headers will still be looked up the same way and hence the current flexibility will be intact. 

c) I don't think adding the `virtual` keyword to functions (which delays the resolution of overridden function from compilation time to runtime) would remove or restrict access to a subset of the information that can be inferred by the IL. My understanding is that the only change that dynamic polymorphism does to the IL is that the IL would not have information to determine which one of the overloaded functions will be called, since that will be resolved at runtime. However, I am not sure what are the core IL generation modules so someone else's comment on that would be more precise.

**Comment by xliang6** [comment]
With the new proposed solution, we will be able to make the API (or extension points) more explicit by using dynamic polymorphism and the keyword `virtual`. With the existing scheme, any function in a class marked as OMR_EXTENSIBLE is implicitly an extension point as it can be overridden in a downstream project. However, to get the correct functional behavior, one would have to make sure we invoke self()->foo() from inside a member function of the same class hierarchy. And this has caused a lot of confusion in the past.

@dibyendumajumdar Can you explain what you mean by "core IL generation modules"?

**Comment by dibyendumajumdar** [comment]
Hi @xliang6 

I understand that there two apis in the Compiler product - the one that is in compiler/il? and the other in compiler/ilgen which is (maybe not wholly) part of the higher level JitBuilder api?

I meant that both the lower level and higher level apis should be available to clients.

Regards

**Comment by samasri** [comment]
# Update about our proposal to change the variability implementation to use dynamic polymorphism

It was agreed upon during the [OMR Compiler Architecture Meeting](https://github.com/eclipse/omr/issues/2571) that after virtualizing all overridden functions in the 20 class hierarchies with the highest number of overridden functions, benchmarks will be performed on OMR. Accordingly, a decision will be made whether to accept the change in the project's variability implementation or not. 

The virtualization process we've used is documented in [a wiki page](https://github.com/samasri/omr/wiki/Virtualization-Process-Guide).

I've virtualized the overridden functions of 4 class hierarchies. I also removed the `self()` keyword from member function calls in those class hierarchies in the OMR project.

# Next steps
* Remove the rest of the `self()` occurrences in the virtualized class hierarchies in OMR. So far, I've only removed the `self()` in member function calls which appear in the form of `self()->`. However, `self()` can appear in function call parameters such as `boo(self())`. I did not handle those cases. This is explained more in the Design Challenge section of this comment.
* Remove all `self()` occurrences in the virtualized class hierarchies in OpenJ9.
* Continue virtualizing the overridden functions in the rest of the 20 class hierarchies.

# Current State
Detailed information about each class hierarchy can be found in [_doc/virtualization_](https://github.com/samasri/omr/tree/master/tools/compiler/OMRStatistics/doc/virtualization). A summary of each class hierarchy's state is found below.

## CodeGenerator
* 199 functions were virtualized
* 3 functions were declared as pure virtual since they had no implementation in the base class.
* `CodeGenerator` constructor was staged into a two-phased initialization. More information about the background for this change can be found in [this issue](https://github.com/eclipse/omr/issues/3235).
* 2 local variables in member functions were renamed (`comp` and `machine` were renamed to `compilation` and `thisMachine` respectively) and a function (`OMR::CodeGenerator::insertInstructionPrefetches()`) was declared to solve consequences of removing `self()` from member function calls ([more info](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/virtualization/CodeGenerator.md#removing-self-from-member-function-calls))
* [OMR PR](https://github.com/eclipse/omr/pull/2639) | [OpenJ9 PR](https://github.com/eclipse/openj9/pull/2087)

## MemoryReference
* 10 functions were virtualized
* 2 functions were ignored for being static
* [OMR PR](https://github.com/eclipse/omr/pull/3162) | [OpenJ9 PR](https://github.com/eclipse/openj9/pull/3606)

## Symbol
* 6 functions were virtualized
* 3 functions were ignored for being static
* 1 problematic function, solved by changing the name of `OMR::AutomaticSymbol::getKind()` in this [PR](https://github.com/eclipse/omr/pull/3241)
* [OMR PR](https://github.com/eclipse/omr/pull/3215) | [OpenJ9 PR](https://github.com/eclipse/openj9/pull/3741)

## Compilation
* 20 functions were virtualized
* 2 functions were ignored for being static
* [OMR PR](https://github.com/eclipse/omr/pull/3226) | [OpenJ9 PR](https://github.com/eclipse/openj9/pull/3796)

## ELFRelocationResolver
* 1 function was virtualized
* Since this class hierarchy is simple and has only 1 overridden function, it was the first class hierarchy whose overridden functions were virtualize; it was used as an example to experiment with virtualizing functions and establish a systematic process
* [OMR PR](https://github.com/eclipse/omr/pull/2584)

# Design Challenge
As mentioned earlier, some of the `self()` functions appear in function call parameters, such as `_machine = new (trHeapMemory()) TR::Machine(self());` ([source](https://github.com/eclipse/omr/blob/71371719e50c740ac33bb0f4c61d459796fe772a/compiler/codegen/OMRCodeGenerator.cpp#L272)). In this case, replacing `self()` with `this` would give a compilation error since the `TR::Machine` constructor expects a `TR::CodeGenerator*` and not `OMR::CodeGenerator*`. This can be solved in two ways:
* Replace `self()` with a static cast to `TR::CodeGenerator`
* Change the `TR::Machine` constructor, which would change the API of OMR.
Please let us know what you think is best. 

@0xdaryl @xliang6 @rwy0717 @nbhuiyan
